### DISCUSSION TITLE
Captcha error at amazon-de pls help me

### DISCUSSION THREAD

**Discussion by vezzkos** [jef/streetmerchant#1316] [Help]
## Expected Behavior

i updatet streetmerchant and get the captcha message.
i try all options. i update my user agent. i try headless mode and nothing helps me.
amazon-de says allwys captcha. i need help pls.

## Steps to Reproduce
1. Update my User Agent
2.Running in Headless Mode
3.Use Vpn
4. Changed my IP
Nothing works for me.

## Environment
Windows 10
**.env file:**

# ** All configuration variables are optional **
# Read https://github.com/jef/streetmerchant#customization for help on customizing this file
#############################################################################################

ASCII_BANNER=""
ASCII_COLOR=""
AUTO_ADD_TO_CART="true"
BROWSER_TRUSTED="false"
COUNTRY="germany"
DESKTOP_NOTIFICATIONS="true"
DISCORD_NOTIFY_GROUP=""
DISCORD_WEB_HOOK=""
EMAIL_PASSWORD=""
EMAIL_TO=""
EMAIL_USERNAME=""
HEADLESS="false"
IN_STOCK_WAIT_TIME="30"
LOG_LEVEL=""
LOW_BANDWIDTH=""
MAX_PRICE_SERIES_3070=""
MAX_PRICE_SERIES_3080=""
MAX_PRICE_SERIES_3090=""
MAX_PRICE_SERIES_CORSAIR_SF=""
MAX_PRICE_SERIES_RX6800=""
MAX_PRICE_SERIES_RX6800XT=""
MAX_PRICE_SERIES_RX6900XT=""
MAX_PRICE_SERIES_RYZEN5600=""
MAX_PRICE_SERIES_RYZEN5800=""
MAX_PRICE_SERIES_RYZEN5900=""
MAX_PRICE_SERIES_RYZEN5950=""
MAX_PRICE_SERIES_SONYPS5C=""
MAX_PRICE_SERIES_SONYPS5DE=""
MICROCENTER_LOCATION=""
MQTT_BROKER_ADDRESS=""
MQTT_BROKER_PORT=""
MQTT_CLIENT_ID=""
MQTT_PASSWORD=""
MQTT_QOS=""
MQTT_TOPIC=""
MQTT_USERNAME=""
NVIDIA_ADD_TO_CART_ATTEMPTS=""
NVIDIA_SESSION_TTL=""
OPEN_BROWSER="true"
PAGE_BACKOFF_MIN=""
PAGE_BACKOFF_MAX=""
PAGE_SLEEP_MIN=""
PAGE_SLEEP_MAX=""
PAGE_TIMEOUT=""
PAGERDUTY_INTEGRATION_KEY=""
PAGERDUTY_SEVERITY=""
PHILIPS_HUE_API_KEY=""
PHILIPS_HUE_CLOUD_ACCESS_TOKEN=""
PHILIPS_HUE_CLOUD_CLIENT_ID=""
PHILIPS_HUE_CLOUD_CLIENT_SECRET=""
PHILIPS_HUE_CLOUD_REFRESH_TOKEN=""
PHILIPS_HUE_LAN_BRIDGE_IP=""
PHILIPS_HUE_LIGHT_COLOR=""
PHILIPS_HUE_LIGHT_IDS=""
PHILIPS_HUE_LIGHT_PATTERN=""
PHONE_CARRIER=""
PHONE_NUMBER=""
PLAY_SOUND="alert.wav"
PROXY_ADDRESS=""
PROXY_PROTOCOL=""
PROXY_PORT=""
PUSHBULLET=""
PUSHOVER_TOKEN=""
PUSHOVER_USER=""
PUSHOVER_PRIORITY=""
SCREENSHOT="true"
SHOW_ONLY_BRANDS=""
SHOW_ONLY_MODELS=""
SHOW_ONLY_SERIES="sonyps5de,sonyps5c"
SLACK_CHANNEL=""
SLACK_TOKEN=""
SMTP_ADDRESS=""
SMTP_PORT=""
STORES="amazon-de,saturn,mediamarkt,saturn,amazon-fr,amazon-es,otto,spielegrotte,smyths,alternate"
TELEGRAM_ACCESS_TOKEN=""
TELEGRAM_CHAT_ID=""
TWILIO_ACCOUNT_SID=""
TWILIO_AUTH_TOKEN=""
TWILIO_FROM_NUMBER=""
TWILIO_TO_NUMBER=""
TWITCH_ACCESS_TOKEN=""
TWITCH_CHANNEL=""
TWITCH_CLIENT_ID=""
TWITCH_CLIENT_SECRET=""
TWITCH_REFRESH_TOKEN=""
TWITTER_ACCESS_TOKEN_KEY=""
TWITTER_ACCESS_TOKEN_SECRET=""
TWITTER_CONSUMER_KEY=""
TWITTER_CONSUMER_SECRET=""
TWITTER_TWEET_TAGS=""
USER_AGENT="Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/87.0.4280.88 Safari/537.36"
WEB_PORT=""

**Comment by vezzkos** [comment]
Here is a picture of the captcha error https://www.bilder-upload.eu/bild-215655-1607595087.png.html

**Comment by SWEETDREAMSIV** [comment]
Try https://jef.codes/streetmerchant/help/troubleshoot/ but for me not work

**Comment by joerg-knitter** [comment]
I also suffer from a lot of captchas introduced recently, and did some extensive tests. In fact, all the captcha troubleshoot hints mentioned in the link above also did not work on my side.
I also changed the PAGE_SLEEP_MIN and PAGE_SLEEP_MAX for amazon-de and amazon-warehouse-de, but I did not succeed. While amazon-de:20000:40000 seemed to work yesterday, even a sleep time of 60s caused captchas today.
As a test, I installed a similar Chrome website check plugin and setup the check of the same Amazon and Amazon Marketplace pages with a 10s interval. For several hours, I did not get any captcha, especially even though it was running side-by-side with streetmerchant. I can even reload the pages every 1-2s manually without any captcha.

My assumption is that you need cookies on amazon.de and/or need to be logged in for avoiding captchas, because when I set PUPPETEER_EXECUTABLE_PATH="C:\Program Files (x86)\Google\Chrome\Application\chrome.exe" and HEADLESS="false", I can see that I am not being logged in and get cookie consent dialogues.
Is there any way to achieve this?

**Comment by vezzkos** [comment]
> Try https://jef.codes/streetmerchant/help/troubleshoot/ but for me not work

i try it but fot me not work too :/

**Comment by SWEETDREAMSIV** [comment]
has anyone tried to change dns? maybe it helps?

**Comment by SWEETDREAMSIV** [comment]
> I also suffer from a lot of captchas introduced recently, and did some extensive tests. In fact, all the captcha troubleshoot hints mentioned in the link above also did not work on my side.
> 
> I also changed the PAGE_SLEEP_MIN and PAGE_SLEEP_MAX for amazon-de and amazon-warehouse-de, but I did not succeed. While amazon-de:20000:40000 seemed to work yesterday, even a sleep time of 60s caused captchas today.
> 
> As a test, I installed a similar Chrome website check plugin and setup the check of the same Amazon and Amazon Marketplace pages with a 10s interval. For several hours, I did not get any captcha, especially even though it was running side-by-side with streetmerchant. I can even reload the pages every 1-2s manually without any captcha.
> 
> 
> 
> My assumption is that you need cookies on amazon.de and/or need to be logged in for avoiding captchas, because when I set PUPPETEER_EXECUTABLE_PATH="C:\Program Files (x86)\Google\Chrome\Application\chrome.exe" and HEADLESS="false", I can see that I am not being logged in and get cookie consent dialogues.
> 
> Is there any way to achieve this?

I also tried when logged in Chrome but it still returns captcha, the bot uses chromium maybe amazon doesn't see you logged in for that.

**Comment by joerg-knitter** [comment]
No, the Chromium as well as the linked "normal" Chrome set in the PUPPETEER_EXECUTABLE_PATH use some kind of "incognito mode", so apparantly no cookies etc. are available in the browser used by streetmerchant. Also when I start the Chromium in the node_modules-puppeteer subfolder, I am not able to make the login persistent.

BTW: This captcha problem indeed also started on my side after updating my source (1/2 week old) to the master yesterday evening. Maybe, this is just a coincidence, but maybe also something has changed that forces this new behaviour.

**Comment by vezzkos** [comment]
mhhh how we can fix that? :(

**Comment by vezzkos** [comment]
for me it is only yesterday. since I made an update. before that everything was cool

**Comment by joerg-knitter** [comment]
@vezzkos: There are extensions e.g. for Chrome that can also check websites on a regular basis. Use these for amazon.de in the meantime until it gets fixed in streetmerchant.

**Comment by Spackler** [comment]
Getting the same for Amazon US

**Comment by vezzkos** [comment]
> @vezzkos: There are extensions e.g. for Chrome that can also check websites on a regular basis. Use these for amazon.de in the meantime until it gets fixed in streetmerchant.

Can you tell me the name of this extension? ^^

**Comment by joerg-knitter** [comment]
Well, before Amazon tries to identify further scripts and extensions and blocks them, I don´t want to name it here literally. Please use Google terms like "chrome extension check websites", and you will find what I mean ;)

**Comment by canopus-2** [comment]
I've found the solution to this problem: just remove the captcha query for amazon stores.

I have included this as fix for the UK stores update here: https://github.com/jef/streetmerchant/pull/1317

**Comment by Spackler** [comment]
Did same to US and fixed that as well. Great find @sohaiiil

**Comment by vezzkos** [comment]
> I have a simple fix for UK stores(just remove the captcha query for amazon stores): #1317

omg it works thank you so much

**Comment by SWEETDREAMSIV** [comment]
> I have a simple fix for UK stores(just remove the captcha query for amazon stores): https://github.com/jef/streetmerchant/pull/1317

I will try as soon as I get home on all the amazon eu stores.

Thank you so much

**Comment by vezzkos** [comment]
> I have a simple fix for UK stores(just remove the captcha query for amazon stores): #1317

my captcha error is gone. but i see the captcha in chromium is that normal?

**Comment by joerg-knitter** [comment]
~~It only partly works: If you start streetmerchant in non-headless mode, you will see that there are a few amazon captchas that are simply ignored now, but I counted maybe 1 out of 6 (with standard PAGE_SLEEP settings) which is at least a better result than before.~~
EDIT: Does not work.

**Comment by vezzkos** [comment]
> It only partly works: If you start streetmerchant in non-headless mode, you will see that there are a few amazon captchas that are simply ignored now, but I counted maybe 1 out of 6 (with standard PAGE_SLEEP settings) which is at least a better result than before.


okay then i can let it run like this and there will still be an alarm if something is in stock? even though I see the captchas?

**Comment by joerg-knitter** [comment]
I would not rely on this. streetmerchant simply makes a text comparison on a loaded website. If it finds "'in den einkaufswagen", it is put to the chart, if it says something different than "Derzeit nicht verfügbar" in the availability text, it opens the browser. But if it opens the captcha page instead, none of those criteria is met.
So when Amazon changes the conditions when captchas are shown, you will not be informed.
The strange thing is, that with the tests this morning, I _only_ got captcha pages. So did Amazon really change the captcha behaviour during the day, or did the captcha detection in streetmerchant reveal itself?
I just tested the "fix" for maybe 1-2 minutes, so I am not sure if I will not get only captchas a few minutes later.

Like mentioned, with one of the Chrome extensions, I did not get a single captcha for hours.

**Comment by vezzkos** [comment]
> I would not rely on this. streetmerchant simply makes a text comparison on a loaded website. If it finds "'in den einkaufswagen", it is put to the chart, if it says something different than "Derzeit nicht verfügbar" in the availability text, it opens the browser. But if it opens the captcha page instead, none of those criteria is met.
> So when Amazon changes the conditions when captchas are shown, you will not be informed.
> The strange thing is, that with the tests this morning, I _only_ got captcha pages. So did Amazon really change the captcha behaviour during the day, or did the captcha detection in streetmerchant reveal itself?
> I just tested the "fix" for maybe 1-2 minutes, so I am not sure if I will not get only captchas a few minutes later.
> 
> Like mentioned, with one of the Chrome extensions, I did not get a single captcha for hours.

the problem is, the chrome extensions only for US & UK i think?. i am from germany :/ if i see mediamarkt and saturn and amazon in chromium headless, i  see everywhere only captcha :/

**Comment by joerg-knitter** [comment]
No, if you use the search tearm suggested above, you will get extensions that allow to take any arbitrary page and select any portion of a page that is then checked for changes in defined intervals. I took the second one with the drop in the logo... ;)
And I use it in a separate standard Chrome desktop browser, not the browser connected with streetmerchant, making it running in parallel to streetmerchant.
And indeed, take care of mediamarkt and saturn regarding captchas - don´t use too low intervals. Same goes especially for GameStop.de, expert and medimax that apparently use IP blockers in case they detect bots.

**Comment by vezzkos** [comment]
> No, if you use the search tearm suggested above, you will get extensions that allow to take any arbitrary page and select any portion of a page that is then checked for changes in defined intervals. I took the second one with the drop in the logo... ;)
> And I use it in a separate standard Chrome desktop browser, not the browser connected with streetmerchant, making it running in parallel to streetmerchant.
> And indeed, take care of mediamarkt and saturn regarding captchas - don´t use too low intervals. Same goes especially for GameStop.de, expert and medimax that apparently use IP blockers in case they detect bots.


mhh also is amazon-de.ts useless i think? or?

**Comment by joerg-knitter** [comment]
Not really: You might use the new proxy list feature that uses different socks5 proxy servers on every request. I haven´t tried this yet and especially don´t have a list of reliable proxy servers.
And if we find a solution on how to save a cookie or log-in in streetmerchant´s browser, the captcha problem on Amazon might be gone.

**Comment by jef** [comment]
I've recently updated the code to use random user agents. This should help with CAPTCHA issues.

**Reply by vezzkos** [reply]
I updatet all but it don´t works for me. :/ Amazon UK and DE allways show CAPTCHA :( 
every now and then it's an hour then captcha comes back. then it goes again for a while. 
In between it works. but then again not.

**Comment by joerg-knitter** [comment]
Works here in non-headless mode for Amazon DE apart from the fact that sometimes the mobile version is being shown.

**Reply by vezzkos** [reply]
so i have it run in normal mode. an hour came just captcha and now everything works again. very funny

**Comment by vezzkos** [comment]
It don´t works. I have Captchas on AmazonDE all time now :((

**Comment by vezzkos** [comment]
can you help me that these captchas are finally gone? amazonDE only shows CAPTCHA ERRORS ;/

**Comment by joerg-knitter** [comment]
I use this:

> ASCII_BANNER=""
ASCII_COLOR=""
AUTO_ADD_TO_CART="true"
BROWSER_TRUSTED=""
COUNTRY=""
DESKTOP_NOTIFICATIONS="true"
DISCORD_NOTIFY_GROUP=""
DISCORD_WEB_HOOK=""
EMAIL_PASSWORD="xyz"
EMAIL_TO="xyz"
EMAIL_USERNAME="xyz@gmail.com"
HEADLESS="false"
IN_STOCK_WAIT_TIME="60"
LOG_LEVEL=""
LOW_BANDWIDTH=""
MAX_PRICE_SERIES_3070=""
MAX_PRICE_SERIES_3080=""
MAX_PRICE_SERIES_3090=""
MAX_PRICE_SERIES_RX6800=""
MAX_PRICE_SERIES_RX6800XT=""
MAX_PRICE_SERIES_RX6900XT=""
MAX_PRICE_SERIES_RYZEN5600=""
MAX_PRICE_SERIES_RYZEN5800=""
MAX_PRICE_SERIES_RYZEN5900=""
MAX_PRICE_SERIES_RYZEN5950=""
MAX_PRICE_CORSAIR_SF=""
MICROCENTER_LOCATION=""
MQTT_BROKER_ADDRESS=""
MQTT_BROKER_PORT=""
MQTT_CLIENT_ID=""
MQTT_PASSWORD=""
MQTT_QOS=""
MQTT_TOPIC=""
MQTT_USERNAME=""
NVIDIA_ADD_TO_CART_ATTEMPTS=""
NVIDIA_SESSION_TTL=""
OPEN_BROWSER=""
PAGE_BACKOFF_MIN=""
PAGE_BACKOFF_MAX=""
PAGE_SLEEP_MIN=""
PAGE_SLEEP_MAX=""
PAGE_TIMEOUT=""
PAGERDUTY_INTEGRATION_KEY=""
PAGERDUTY_SEVERITY=""
PHILIPS_HUE_API_KEY=""
PHILIPS_HUE_CLOUD_ACCESS_TOKEN=""
PHILIPS_HUE_CLOUD_CLIENT_ID=""
PHILIPS_HUE_CLOUD_CLIENT_SECRET=""
PHILIPS_HUE_CLOUD_REFRESH_TOKEN=""
PHILIPS_HUE_LAN_BRIDGE_IP=""
PHILIPS_HUE_LIGHT_COLOR=""
PHILIPS_HUE_LIGHT_IDS=""
PHILIPS_HUE_LIGHT_PATTERN=""
PHONE_CARRIER=""
PHONE_NUMBER=""
PLAY_SOUND="alarm.ogg"
PROXY_ADDRESS=""
PROXY_PROTOCOL=""
PROXY_PORT=""
PUSHBULLET=""
PUSHOVER_TOKEN=""
PUSHOVER_USER=""
PUSHOVER_PRIORITY=""
SCREENSHOT=""
SHOW_ONLY_BRANDS="sony"
SHOW_ONLY_MODELS="ps5 console"
SHOW_ONLY_SERIES="sonyps5c"
SLACK_CHANNEL=""
SLACK_TOKEN=""
SMTP_ADDRESS="smtp.gmail.com"
SMTP_PORT="465"
STORES="xyz"
TELEGRAM_ACCESS_TOKEN=""
TELEGRAM_CHAT_ID=""
TWILIO_ACCOUNT_SID=""
TWILIO_AUTH_TOKEN=""
TWILIO_FROM_NUMBER=""
TWILIO_TO_NUMBER=""
TWITCH_ACCESS_TOKEN=""
TWITCH_CHANNEL=""
TWITCH_CLIENT_ID=""
TWITCH_CLIENT_SECRET=""
TWITCH_REFRESH_TOKEN=""
TWITTER_ACCESS_TOKEN_KEY=""
TWITTER_ACCESS_TOKEN_SECRET=""
TWITTER_CONSUMER_KEY=""
TWITTER_CONSUMER_SECRET=""
TWITTER_TWEET_TAGS=""
WEB_PORT=""

**Reply by vezzkos** [reply]
Many Thanks. yes I have all of that too. i have updated everything. everything is up to date. but also with the random user agents I only have captchas at AmazonDE and UK in normal mode but also in headless mode. the captcha doesn't go away anymore. am desperate. :(

**Comment by joerg-knitter** [comment]
Maybe you are kind of blocked, did you reconnect with a new/another VPN connection to get a different IP address?
Make a test by visiting those pages directly - you get the URLs from the amazon-de.ts file in the subfolder \src\store\model.
How may stores have you added? Maybe a delay might help, like amazon:10000:20000.

**Reply by jef** [reply]
I would try this @vezzkos, it's good advice. Definitely try the pages in your normal browser and see if they work.

Take a look at [Option 2](https://jef.codes/streetmerchant/help/troubleshoot/#option-2) here as well -- that could also help.

**Reply by vezzkos** [reply]
I do not get it. what exactly i have to do with option 2. can you tell me what exactly i have to do so that it works in the chrome browser please?

**Comment by gigi2006** [comment]
so i must say that since the random user and vpn i am totally happy.
Although I still used on a VM my own user agents etc. and there it also works. 

but have very few captchas.

only one problem I still have but.
as soon as times captchas come and I want to solve, it is impossible. because all the other pages are loaded too quickly and the captcha always goes to the back. so I can just never solve it.

**Reply by jef** [reply]
This will probably come in the future, but as now, doesn't exists.

**Reply by vezzkos** [reply]
Did you know a good VPN Mediamarkt and Saturn allways blocked me with 403 error. i use Proton VPN but all ips instantly blocked by Mediamarkt and Saturn. But this captcha problem with amazon is solved. thank you so much.

**Comment by joerg-knitter** [comment]
I have no idea about VPN services, but searched for an overview and found the German page vpn-anbieter-vergleich-test. I don´t know if the information there is just general promotion, but since the author also has a YouTube channel for several years, he would get very negative feedback on the long term if this is just and only an advertisment service (even though he has some affiliate links).
From the list of VPNs, I finally chose one that had a 3-year subscription at a very low price with good results, and the pages mentioned work here testing with several VPN locations, DE included. 
However, sometimes you have to disconnect and reconnect again if one of the shops make problems - and if this happens more often on a certain shop, make sure to change the timings. 
However, I removed MediaMarkt and Saturn because I don´t trust them (anymore) regarding the catastrophic delivery situation of PS5 consoles, but it´s your decision ;)...

P.S.: Sorry for not going into details, but I want to avoid that the shops get further attention and start blocking tools and services...

**Comment by jef** [comment]
I've also updated the UAs. Feel free to pull the latest and test that out.

Thanks!
